app/scripts/modules/core/src/search/global/GlobalSearch.tsx

```diff
@@ -0,0 +1,405 @@
+    const refToFocus = flattenedRefs[flattenedRefs.length - 1];
+    refToFocus && refToFocus.focus();
+  }
+
+  private queryChanged({ target }: React.ChangeEvent<HTMLInputElement>) {
+    this.setState({ query: target.value }, () => {
+      if (this.state.query.length >= 3) {
+        this.executeQuery();
+      } else {
+        this.considerMinLengthWarning();
+      }
+    });
+  }
+
+  private navigateResult(event: React.KeyboardEvent<HTMLAnchorElement>) {
+    const { which, target } = event;
+    if (which === 27) { // escape
+      ReactGA.event({ category: 'Global Search', action: 'Keyboard Nav', label: 'escape (from result)' });
+      this.reset();
+    }
+    if (which === 9) { // tab - let it navigate automatically, but close menu if on the last result
+      const flattenedRefs = flatten(this.resultRefs);
+      const lastResultRef = flattenedRefs[flattenedRefs.length - 1];
+      if (target === lastResultRef) {
+        ReactGA.event({ category: 'Global Search', action: 'Keyboard Nav', label: 'tab (from result)' });
+        this.hideResults();
+        return;
+      }
+    }
+    if (which === 40) { // down
+      ReactGA.event({ category: 'Global Search', action: 'Keyboard Nav', label: 'down (from result)' });
+      const flattenedRefs = flatten(this.resultRefs);
+      const currentRefIndex = flattenedRefs.indexOf((target as HTMLElement));
+      const nextResultRef = flattenedRefs[currentRefIndex + 1];
+
+      nextResultRef && nextResultRef.focus();
+      event.preventDefault();
+    }
+    if (which === 38) { // up
+      ReactGA.event({ category: 'Global Search', action: 'Keyboard Nav', label: 'up (from result)' });
+      const flattenedRefs = flatten(this.resultRefs);
+      const currentRefIndex = flattenedRefs.indexOf((target as HTMLElement));
+      const prevResultRef = flattenedRefs[currentRefIndex - 1];
+
+      prevResultRef && prevResultRef.focus();
+      event.preventDefault();
+    }
+  }
+
+  @Debounce(200)
+  private executeQuery() {
+    const { infrastructureSearchService } = ReactInjector;
+    const search = infrastructureSearchService.getSearcher();
+
+    ReactGA.event({ category: 'Global Search', action: 'Query', label: this.state.query });
+    this.setState({ querying: true, showMinLengthWarning: false });
+
+    search.query(this.state.query).then((result) => {
+      const showMinLengthWarning = !!this.state.query && this.state.query.length < 3;
+
+      const categories = result
+        .filter(({ results }) => results.length)
+        .map((category) => ({
+          ...category,
+          results: searchRank(category.results).slice(0, 5)
+        }))
+        .sort((a, b) => a.order - b.order);
+
+      this.resultRefs = range(categories.length).map(() => []);
+
+      this.setState({
+        querying: false,
+        showSearchResults: !!this.state.query && !showMinLengthWarning,
+        categories,
+        showMinLengthWarning
+      });
+    });
+  }
+
+  @Debounce(200)
```

Comment:
showing this immediately ended up being a little jarring as you were typing in a search term, so this leaves a little breathing room to check whether you're actually trying to search for < 3 chars or if you're just in the middle of typing.

Reply:
I think we've used 300ms as the debounce rate elsewhere

Reply:
think it's worth switching to 300 for the executeQuery call too? I was just porting it over verbatim.

Reply:
Yeah, I think so (I see where you're getting the 200 from now, too)